# Exercise 02 -  OLAP Cubes - Roll Up and Drill Down

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [1]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'abbas'
DB_PASSWORD = 'root'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://abbas:root@127.0.0.1:5432/pagila_star


'Connected: abbas@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

## Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [3]:
%%time
%%sql

SELECT dimDate.day, dimMovie.rating, dimCustomer.country, sum(sales_amount) as revenue
FROM factSales
JOIN dimMovie ON (dimMovie.movie_key = factSales.movie_key)
JOIN dimDate ON (dimDate.date_key = factSales.date_key)
JOIN dimCustomer ON (dimCustomer.customer_key = factSales.customer_key)
GROUP BY (dimDate.day, dimCustomer.country, dimMovie.rating)
ORDER BY revenue DESC 
LIMIT 20;

 * postgresql://abbas:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 1.89 ms, sys: 3.52 ms, total: 5.41 ms
Wall time: 189 ms


[(30, 'PG', 'India', Decimal('215.52')),
 (18, 'PG-13', 'India', Decimal('208.58')),
 (30, 'R', 'India', Decimal('194.56')),
 (30, 'G', 'China', Decimal('193.63')),
 (30, 'PG-13', 'China', Decimal('185.59')),
 (30, 'PG-13', 'India', Decimal('178.57')),
 (21, 'R', 'India', Decimal('177.61')),
 (17, 'PG-13', 'China', Decimal('176.63')),
 (30, 'R', 'China', Decimal('170.58')),
 (1, 'R', 'India', Decimal('164.64')),
 (18, 'NC-17', 'India', Decimal('162.65')),
 (30, 'NC-17', 'India', Decimal('162.61')),
 (18, 'PG-13', 'China', Decimal('160.62')),
 (18, 'PG', 'India', Decimal('158.63')),
 (21, 'PG-13', 'India', Decimal('157.61')),
 (30, 'G', 'India', Decimal('156.61')),
 (21, 'NC-17', 'China', Decimal('152.68')),
 (27, 'PG-13', 'India', Decimal('151.69')),
 (30, 'PG', 'China', Decimal('148.66')),
 (20, 'PG-13', 'India', Decimal('147.69'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

## Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [4]:
%%time
%%sql

SELECT dimDate.day, dimMovie.rating, dimCustomer.district, sum(sales_amount) as revenue
FROM factSales
JOIN dimMovie ON (dimMovie.movie_key = factSales.movie_key)
JOIN dimDate ON (dimDate.date_key = factSales.date_key)
JOIN dimCustomer ON (dimCustomer.customer_key = factSales.customer_key)
GROUP BY (dimDate.day, dimCustomer.district, dimMovie.rating)
ORDER BY revenue DESC 
LIMIT 20;

 * postgresql://abbas:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 1.97 ms, sys: 2.37 ms, total: 4.34 ms
Wall time: 231 ms


[(18, 'PG-13', 'West Bengali', Decimal('56.90')),
 (30, 'PG-13', 'Southern Tagalog', Decimal('53.88')),
 (30, 'G', 'Inner Mongolia', Decimal('52.91')),
 (30, 'PG-13', 'Buenos Aires', Decimal('52.89')),
 (27, 'NC-17', 'Buenos Aires', Decimal('49.86')),
 (18, 'NC-17', 'So Paulo', Decimal('48.90')),
 (20, 'PG', 'Shandong', Decimal('47.89')),
 (1, 'PG', 'California', Decimal('44.92')),
 (19, 'G', 'Shandong', Decimal('43.92')),
 (21, 'R', 'So Paulo', Decimal('43.90')),
 (17, 'PG-13', 'Shandong', Decimal('42.94')),
 (19, 'NC-17', 'California', Decimal('42.93')),
 (2, 'R', 'Buenos Aires', Decimal('41.92')),
 (30, 'PG-13', 'West Bengali', Decimal('40.89')),
 (19, 'G', 'California', Decimal('39.94')),
 (28, 'PG-13', 'England', Decimal('39.92')),
 (27, 'NC-17', 'California', Decimal('39.92')),
 (21, 'R', 'Uttar Pradesh', Decimal('39.91')),
 (18, 'PG-13', 'Maharashtra', Decimal('38.94')),
 (12, 'NC-17', 'Buenos Aires', Decimal('38.90'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>